src/VisualStudio/Core/Def/Implementation/Progression/GraphQueryManager.cs

```diff
@@ -152,33 +156,45 @@ private static bool IsTrackingContext(WeakReference<IGraphContext> weakContext)
-            return whenAllTask.SafeContinueWith(
-                t => ProcessGraphTasks(t.Result, context),
-                cancellationToken, TaskContinuationOptions.None, TaskScheduler.Default);
-        }
-
-        private static void ProcessGraphTasks(GraphBuilder[] graphBuilders, IGraphContext context)
+        private static async Task PopulateContextGraphAsync(Solution solution, List<IGraphQuery> graphQueries, IGraphContext context)
         {
-            // Perform the actual graph transaction 
-            using var transaction = new GraphTransactionScope();
-            // Remove any links that may have been added by a previous population. We don't
-            // remove nodes to maintain node identity, matching the behavior of the old
-            // providers.
-            context.Graph.Links.Clear();
-
-            foreach (var graphBuilder in graphBuilders)
+            try
             {
-                graphBuilder.ApplyToGraph(context.Graph);
+                var cancellationToken = context.CancelToken;
+                var graphBuilderTasks = graphQueries.Select(q => q.GetGraphAsync(solution, context, cancellationToken)).ToArray();
+                var graphBuilders = await Task.WhenAll(graphBuilderTasks).ConfigureAwait(false);
 
-                context.OutputNodes.AddAll(graphBuilder.CreatedNodes);
+                // Perform the actual graph transaction 
+                using var transaction = new GraphTransactionScope();
+
+                // Remove any links that may have been added by a previous population. We don't
+                // remove nodes to maintain node identity, matching the behavior of the old
+                // providers.
+                context.Graph.Links.Clear();
+
+                foreach (var graphBuilder in graphBuilders)
+                {
+                    graphBuilder.ApplyToGraph(context.Graph);
+
+                    context.OutputNodes.AddAll(graphBuilder.CreatedNodes);
+                }
+
+                transaction.Complete();
             }
+            catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceled(ex))
+            {
+                var errorReportingService = solution.Workspace.Services.GetRequiredService<IErrorReportingService>();
+
+                // The error message shown is confusing as it's not necessarily an OOP process, but unfortunately we are past the localization deadline so we must use
+                // the strings we already have.
+                errorReportingService.ShowErrorInfoInActiveView(
+                    ServicesVSResources.Unfortunately_a_process_used_by_Visual_Studio_has_encountered_an_unrecoverable_error_We_recommend_saving_your_work_and_then_closing_and_restarting_Visual_Studio,
+                    new InfoBarUI(WorkspacesResources.Show_Stack_Trace, InfoBarUI.UIKind.HyperLink, () =>
+                            errorReportingService.ShowDetailedErrorInfo(ex), closeAfterAction: true));
 
```

Comment:
Isn't the gold info-bar displayed when SoftCrashException is thrown?

Reply:
If it is, I couldn't find the code.

Reply:
And I *did* look a bit, but I'll be honest the flow of those things wasn't entirely obvious to me. This also implicitly covers the case too where it's not a SoftCrashException but any other exception that happens while processing a graph query which might just be useful in general.

Reply:
[Here](http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Shared/RemoteEndPoint.cs,405) we call `UnexpectedExceptionThown`, which shows the infor bar: http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Remote/JsonRpcConnection.cs,2279bdfdaecd12bb,references

Reply:
So what should I do then? I still want to show it for local exceptions as well.

When testing this I was testing with exceptions being thrown on the local side; I guess I should have injected a fault on the OOP side and tested that too.

Reply:
Are we changing some state while this code is calculating the graph? If not then the potential for corruption is limited and can just return empty results (and report NFW).
If we want to do more for 16.6 we should show node in the graph that indicates the failure.

Reply:
It's hard to say: to ultimately report data this is calling into progression APIs which are semi-mutable -- if we're reporting something more than once we have to update existing mutable state with the new computation. This also could break more than just solution explorer since technically others _could_ use this API too.

But I think you're making me think that yes if the remote case is already covered then I'll just remove the info bar, but leave the NFW code. That's fine with me.

Reply:
Changed and confirmed the gold bar does appear if the remote throws an exception.